packages/dara-core/dara/core/internal/encoder_registry.py

```diff
@@ -134,7 +143,7 @@ def _get_pandas_array_encoder(array_type: Type[Any], dtype: Any, raise_: bool =
     pandas.Index: Encoder(serialize=lambda x: x.to_list(), deserialize=lambda x: pandas.Index(x)),
     pandas.Timestamp: Encoder(serialize=lambda x: x.isoformat(), deserialize=lambda x: pandas.Timestamp(x)),
     pandas.DataFrame: Encoder(
-        serialize=lambda x: jsonable_encoder(x.to_dict(orient='dict')),
+        serialize=lambda x: jsonable_encoder(_tuple_key_serialize(x.to_dict(orient='dict'))),
         deserialize=lambda x: x if isinstance(x, pandas.DataFrame) else _not_implemented(x, pandas.DataFrame),
```

Comment:
Can we modify the default deserializer to detect the `__tuple__` prefixed keys and automatically turn them into a multi index or is that not possible without knowing the types? We could encode more information in the serialized tuple if need be

Reply:
I guess I can add a deserializer for df. If a user add an annotation type DataFrame, I can check if the value is a dict type, if it is, turn the dict into a multi indexed df.

However, if a user doesn't add an annotation type, or the type is List[DataFrame]/Dict[str,DataFrame], the deserializer won't work.

Reply:
Yeah the deserializer *only* runs when the user annotates it as a DF, which I think is fine.